impy/kinematics.py

```diff
@@ -144,17 +251,21 @@ def __init__(
         beam=None,
         nuc1_prop=None,
         nuc2_prop=None,
+        particle1=None,
+        particle2=None,
     ):
-
+        self._init_guard()
         # Catch input errors
         assert (
             np.sum(np.asarray([ecm, plab, elab, ekin, beam], dtype="bool")) == 1
         ), "Define exclusively one energy/momentum definition"
-        assert p1pdg or nuc1_prop, (
-            "Define either particle id or " + "nuclear properties for side 1."
+        assert np.sum(np.asarray([p1pdg, nuc1_prop, particle1], dtype="bool")) == 1, (
+            "Define either particle id, "
+            + "nuclear properties or particle1 for side 1."
         )
-        assert p2pdg or nuc2_prop, (
-            "Define either particle id or " + "nuclear properties for side 2."
+        assert np.sum(np.asarray([p2pdg, nuc2_prop, particle2], dtype="bool")), (
+            "Define either particle id, "
+            + "nuclear properties or particle2 for side 2."
         )
```

Comment:
It took me a while to understand this myself, but `assert` is more an executable comment. It documents that the class has certain invariants (that's the jargon) which are upheld. For example, if you made a class that represents only positive numbers, then an assert at the end of the `__init__` could document this.
```
class PositiveNumber:
    def __init__(self, value):
          if value <= 0: raise ValueError("only positive numbers allowed")
          self._value = value
          assert self._value > 0
```
The assert looks superfluous, but if you think of it like a comment to another programmer, it makes sense. Especially in a non-trivial piece of code. If for some reason the input validation does not work as intended (a bug), you will trigger the assert instead of getting some other confusing error downstream in the code.

Reply:
All `assert`s are changed to exceptions.